# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: docshell/base/BrowsingContext.cpp
# Commit: cefac6461948
# Full Hash: cefac6461948d6b8ef6cd6b1f563a2e487e0bbac
# Author: Kris Maglione <maglione.k@gmail.com>
# Date: 2020-02-07 03:55:51
# Regressor Bug: 1582832
# File Overlap Count: 1
# Description:
#   Bug 1582832: Part 2 - Don't attach a BrowsingContext before a FrameLoader is initialized. r=nika
#   
#   There are cases where the initial BrowsingContext that we create for a
#   FrameLoader will never have a DocShell created for it. In the particular case
#   when the frame it belongs to is part of an inactive document, trying to
# ==============================================================================

diff -r 6fd57b83bb52 -r cefac6461948 docshell/base/BrowsingContext.cpp
--- a/docshell/base/BrowsingContext.cpp	Thu Feb 06 19:07:56 2020 +0000
+++ b/docshell/base/BrowsingContext.cpp	Thu Feb 06 19:08:04 2020 +0000
@@ -126,7 +126,7 @@
 }
 
 /* static */
-already_AddRefed<BrowsingContext> BrowsingContext::Create(
+already_AddRefed<BrowsingContext> BrowsingContext::CreateDetached(
     BrowsingContext* aParent, BrowsingContext* aOpener, const nsAString& aName,
     Type aType) {
   MOZ_DIAGNOSTIC_ASSERT(!aParent || aParent->mType == aType);
@@ -197,12 +197,24 @@
 
   context->mFields.SetWithoutSyncing<IDX_IsActive>(true);
 
-  Register(context);
+  return context.forget();
+}
 
-  // Attach the browsing context to the tree.
-  context->Attach();
+already_AddRefed<BrowsingContext> BrowsingContext::Create(
+    BrowsingContext* aParent, BrowsingContext* aOpener, const nsAString& aName,
+    Type aType) {
+  RefPtr<BrowsingContext> bc(CreateDetached(aParent, aOpener, aName, aType));
+  bc->EnsureAttached();
+  return bc.forget();
+}
 
-  return context.forget();
+void BrowsingContext::EnsureAttached() {
+  if (!mEverAttached) {
+    Register(this);
+
+    // Attach the browsing context to the tree.
+    Attach();
+  }
 }
 
 /* static */
@@ -242,6 +254,10 @@
 
   // Caller handles attaching us to the tree.
 
+  if (aInit.mCached) {
+    context->mEverAttached = true;
+  }
+
   return context.forget();
 }
 
@@ -254,6 +270,7 @@
       mBrowsingContextId(aBrowsingContextId),
       mGroup(aGroup),
       mParent(aParent),
+      mEverAttached(false),
       mIsInProcess(false),
       mIsDiscarded(false),
       mDanglingRemoteOuterProxies(false),
@@ -266,6 +283,7 @@
 void BrowsingContext::SetDocShell(nsIDocShell* aDocShell) {
   // XXX(nika): We should communicate that we are now an active BrowsingContext
   // process to the parent & do other validation here.
+  MOZ_DIAGNOSTIC_ASSERT(mEverAttached);
   MOZ_RELEASE_ASSERT(aDocShell->GetBrowsingContext() == this);
   mDocShell = aDocShell;
   mDanglingRemoteOuterProxies = !mIsInProcess;
@@ -336,6 +354,9 @@
 }
 
 void BrowsingContext::Attach(bool aFromIPC) {
+  MOZ_DIAGNOSTIC_ASSERT(!mEverAttached);
+  mEverAttached = true;
+
   MOZ_LOG(GetLog(), LogLevel::Debug,
           ("%s: Connecting 0x%08" PRIx64 " to 0x%08" PRIx64,
            XRE_IsParentProcess() ? "Parent" : "Child", Id(),
@@ -369,6 +390,8 @@
 }
 
 void BrowsingContext::Detach(bool aFromIPC) {
+  MOZ_DIAGNOSTIC_ASSERT(mEverAttached);
+
   MOZ_LOG(GetLog(), LogLevel::Debug,
           ("%s: Detaching 0x%08" PRIx64 " from 0x%08" PRIx64,
            XRE_IsParentProcess() ? "Parent" : "Child", Id(),
@@ -760,6 +783,7 @@
 bool BrowsingContext::WriteStructuredClone(JSContext* aCx,
                                            JSStructuredCloneWriter* aWriter,
                                            StructuredCloneHolder* aHolder) {
+  MOZ_DIAGNOSTIC_ASSERT(mEverAttached);
   return (JS_WriteUint32Pair(aWriter, SCTAG_DOM_BROWSING_CONTEXT, 0) &&
           JS_WriteUint32Pair(aWriter, uint32_t(Id()), uint32_t(Id() >> 32)));
 }
@@ -1204,6 +1228,7 @@
 }
 
 BrowsingContext::IPCInitializer BrowsingContext::GetIPCInitializer() {
+  MOZ_DIAGNOSTIC_ASSERT(mEverAttached);
   MOZ_DIAGNOSTIC_ASSERT(mType == Type::Content);
 
   IPCInitializer init;
@@ -1462,6 +1487,7 @@
 
 void IPDLParamTraits<dom::BrowsingContext*>::Write(
     IPC::Message* aMsg, IProtocol* aActor, dom::BrowsingContext* aParam) {
+  MOZ_DIAGNOSTIC_ASSERT(!aParam || aParam->EverAttached());
   uint64_t id = aParam ? aParam->Id() : 0;
   WriteIPDLParam(aMsg, aActor, id);
   if (!aParam) {